src/modules/vmount/input_mavlink.cpp

```diff
@@ -85,99 +77,98 @@ int InputMavlinkROI::initialize()
 	return 0;
 }
 
-int InputMavlinkROI::update_impl(unsigned int timeout_ms, ControlData **control_data, bool already_active)
+InputMavlinkROI::UpdateResult
+InputMavlinkROI::update(unsigned int timeout_ms, ControlData &control_data, bool already_active)
 {
-	// already_active is unused, we don't care what happened previously.
-
-	// Default to no change, set if we receive anything.
-	*control_data = nullptr;
-
 	const int num_poll = 2;
 	px4_pollfd_struct_t polls[num_poll];
-	polls[0].fd = 		_vehicle_roi_sub;
-	polls[0].events = 	POLLIN;
-	polls[1].fd = 		_position_setpoint_triplet_sub;
-	polls[1].events = 	POLLIN;
+	polls[0].fd = _vehicle_roi_sub;
+	polls[0].events = POLLIN;
+	polls[1].fd = _position_setpoint_triplet_sub;
+	polls[1].events = POLLIN;
 
 	int ret = px4_poll(polls, num_poll, timeout_ms);
 
-	if (ret < 0) {
-		return -errno;
+	if (ret <= 0) {
+		return UpdateResult::NoUpdate;
 	}
 
-	if (ret == 0) {
-		// Timeout, _control_data is already null
+	if (polls[0].revents & POLLIN) {
+		vehicle_roi_s vehicle_roi;
+		orb_copy(ORB_ID(vehicle_roi), _vehicle_roi_sub, &vehicle_roi);
 
-	} else {
-		if (polls[0].revents & POLLIN) {
-			vehicle_roi_s vehicle_roi;
-			orb_copy(ORB_ID(vehicle_roi), _vehicle_roi_sub, &vehicle_roi);
+		if (vehicle_roi.mode == vehicle_roi_s::ROI_NONE) {
 
-			_control_data.gimbal_shutter_retract = false;
+			control_data.type = ControlData::Type::Neutral;
+			_cur_roi_mode = vehicle_roi.mode;
+			control_data.sysid_primary_control = _parameters.mav_sysid;
+			control_data.compid_primary_control = _parameters.mav_compid;
 
-			if (vehicle_roi.mode == vehicle_roi_s::ROI_NONE) {
+			return UpdateResult::UpdatedActiveOnce;
 
-				_control_data.type = ControlData::Type::Neutral;
-				*control_data = &_control_data;
-				_cur_roi_mode = vehicle_roi.mode;
+		} else if (vehicle_roi.mode == vehicle_roi_s::ROI_WPNEXT) {
+			control_data.type = ControlData::Type::LonLat;
+			_read_control_data_from_position_setpoint_sub(control_data);
+			// FIXME: where does this number come from?
```

Comment:
Check the documentation for `pitch_fixed_angle`

Reply:
Ah, it's just a magic number to signal "ignore". Should we use NAN instead?

Reply:
That works too.